File: checkers/go/sql_injection.test.go
package main

import (
	"database/sql"
	"fmt"
)

// Vulnerable: fmt.Sprintf with SQL query
func vulnerableQuery(db *sql.DB, userID string) error {
	// <expect-error>
	query := fmt.Sprintf("SELECT * FROM users WHERE id = %s", userID)
	_, err := db.Query(query)
	return err
}

// Vulnerable: String concatenation
func vulnerableConcat(db *sql.DB, name string) error {
	// <expect-error>
	query := "SELECT * FROM users WHERE name = '" + name + "'"
	_, err := db.Query(query)
	return err
}

// Vulnerable: fmt.Sprintf with INSERT
func vulnerableInsert(db *sql.DB, name, email string) error {
	// <expect-error>
	query := fmt.Sprintf("INSERT INTO users (name, email) VALUES ('%s', '%s')", name, email)
	_, err := db.Exec(query)
	return err
}

// <no-error> - Parameterized query with $1
func secureQueryPostgres(db *sql.DB, userID string) error {
	query := "SELECT * FROM users WHERE id = $1"
	_, err := db.QueryRow(query, userID)
	return err
}

// <no-error> - Parameterized query with ?
func secureQueryMySQL(db *sql.DB, userID string) error {
	query := "SELECT * FROM users WHERE id = ?"
	_, err := db.QueryRow(query, userID)
	return err
}

// <no-error> - Prepared statement
func securePrepared(db *sql.DB, userID string) error {
	stmt, _ := db.Prepare("SELECT * FROM users WHERE id = $1")
	defer stmt.Close()
	_, err := stmt.Query(userID)
	return err
}

File: checkers/php/sql_injection.test.php
<?php
// Vulnerable: mysqli->query with interpolated string
$id = $_GET['id'];
// <expect-error>
$result = $mysqli->query("SELECT * FROM users WHERE id = $id");

// Vulnerable: concatenation in query
$username = $_POST['username'];
// <expect-error>
$conn->query("SELECT * FROM users WHERE username = '$username'");

// Vulnerable: mysql_query with variable
// <expect-error>
mysql_query("SELECT * FROM products WHERE name = '$name'");

// Vulnerable: mysqli_query function
// <expect-error>
mysqli_query($db, "DELETE FROM users WHERE id = $id");

// Vulnerable: PDO query with interpolation
// <expect-error>
$pdo->query("SELECT * FROM orders WHERE user_id = $userId");

// <no-error> - Prepared statement with mysqli
$stmt = $mysqli->prepare("SELECT * FROM users WHERE id = ?");
$stmt->bind_param("i", $id);
$stmt->execute();

// <no-error> - Prepared statement with PDO
$stmt = $pdo->prepare("SELECT * FROM users WHERE id = :id");
$stmt->execute(['id' => $id]);

// <no-error> - Static query
$result = $mysqli->query("SELECT COUNT(*) FROM users");

// <no-error> - Query with constant
$result = $db->query(USERS_QUERY);
?>
